lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -26,16 +26,20 @@ def __init__(self, context):
     def execute(self) -> int:
         errors = 0
 
-        groups = [group for group in self.context.group.split(",") if group]
+        groups = self.context.group.split(",")
         # groups is never empty and if the user doesn't specify a --group,
         # then it has the value "default"
         for group in groups:
-            if self.verify_tool_group(group) != 0:
+            if not group:
+                self.logger.warn("Tool group name can not be empty")
+                errors = 1
```

Comment:
Semantically, connecting a "warning" with `errors = 1` bothers me. And, in the end, the UNIX command convention of success == 0 or failure code means that clearly communicating "warnings" (or "partial" or "qualified" success) is awkward. Personally, I think that if we're going to continue but just "warn" about the ignored terms, the command should succeed, not fail; or a script using it is going to fail anyway and we might as well have terminated here.

Reply:
Actually, I need to refine my previous comment a bit.  There are types of failures, which we're familiar with seeing embedded in reporting systems:  critical/fatal, error, warning, informational, etc.  If we encounter a critical/fatal error (such as a missing config file or a missing run directory), then we issue a diagnostic and stop immediately with an error exit status.  For warnings and informationals, we continue.  For "error-level" problems, we are adopting a policy (for this tool, at least), that we will issue a diagnostic and continue operation.  And, I'm suggesting that the final exit status should reflect whether there were any successes among the errors.

Reply:
> I'm willing to go along with the idea that, if we clear any item, then we report success

I'd suggest instead that we report success if we successfully complete everything we're asked to do. In theory I'd agree that if we do anything successfully it's a "partial success" but that's simply not possible to express at the command level and think that, on the whole, it's probably safer to treat partial success as failure than as unqualified success.

This case is a bit different, because the interspersed blank values aren't telling us to do anything. We can choose to treat these "non-commands" as errors, but that seems unnecessarily strict; I'd rather treat them as no-ops which don't affect the overall status.

Reply:
I'd be tempted to say we can defer this to a separate PR, but I'm not sure it's really disconnected since this PR has added the syntactic rope (the comma-separated list) to lower oneself into this semantic trap. I'd rather build the ladder at the same time.

(My daughters got my mother into playing The Sims; a very unconventionally creative computer user, my mother figured out that the best way to remove a sim she didn't want anymore was to get it into a swimming pool and delete the ladder ...)

Reply:
I guess to keep it simple for now we can treat empty values as non error case. Client asked us to remove an empty value and we did remove an empty value (but we shouldn't tell them that all we did was no-op except the warning message) because in theory removing an empty value option and not doing anything with empty value have the same effect :smile: but the client is happy.

Reply:
> User asked us to remove an empty value

That's not necessarily true:  the command invocation, issued (e.g.) by the User's _tool_, requested that we remove an empty value...but I'm pretty sure that the User didn't want that....

> in theory removing an empty value option and not doing anything with empty value have the same effect 😄 but the user is happy.

Which empty value did the our command remove?  How do we know that we removed the one that the user wanted us to remove?  😀  Seriously, though, if _I_ were the user, I would not be happy, because I wouldn't have specified an empty value...and, if one somehow _got_ specified, then it's probably  because the value that I **wanted** to have there somehow got omitted.

I'm inclined to follow Dave's suggestion that "partial success" be deemed an error, and that blank options at best create a partial success situation.

Reply:
> That's not necessarily true: the command invocation, issued (e.g.) by the User's _tool_, requested that we remove an empty value...but I'm pretty sure that the User didn't want that....
> 
> >

Yes the user didn't want to, but the computer (command) don't know that, it will think that the user passed an empty value. That's what I was referring to when I said "us" :stuck_out_tongue_closed_eyes: Command doesn't care how did the user passed an empty value, all it cares is whether it got an empty value.

But anyway +1 from my side for partial success to be deemed as an error, if that is the case as the code stands right now, I don't think I need to update the error status.

Reply:
This seems to have dropped my final paragraph,

> This case is a bit different, because the interspersed blank values aren't telling us to do anything. We can choose to treat these "non-commands" as errors, but that seems unnecessarily strict; I'd rather treat them as no-ops which don't affect the overall status.

Yes, I think we need to "fail" (non-zero return status) on partial success because the CLI conventions give us no real choice; but as long as there's at least one non-blank value in each option argument list (and operating on it succeeds) I'd prefer not to *fail* due to an interspersed blank value. We can issue a warning to `stderr`, but I think I would prefer that it not generate a non-zero return value. Failing on `--group=x,,y` seems unnecessarily harsh (as well as somewhat misleading) if acting on `x` and `x` both succeed.